### PR TITLE
WIP: feat: SP builders in Core

### PR DESCRIPTION
**Pull request by toyamaza**
This PR moves SingleHitSpacePointBuilder and DoubleHitSpacePointBuilder from Plugins/Digitization to Core in order to use them in athena.

The previous SP builders used PlanarModuleCluster as input. The segmentation of the strip is needed for the double space point builder, so I'm adding a simple ModuleCluster class that contains Measurement and segment in Examples, but it's open for discussion.

The calculation of the variances in R and Z is introduced as they are required in seeding.

The strip SP builder is not supported in ActsExamples yet. I think the strip digitization needs to be added.

### CODE REVIEW

File: Core/src/Digitization/CartesianSegmentation.cpp [hunk 1]
```diff
@@ -10,7 +10,7 @@
 // CartesianSegmentation.cpp, Acts project
 ///////////////////////////////////////////////////////////////////
```
Comment (asalzburger) [inline Core/src/Digitization/CartesianSegmentation.cpp:11]: These headers can be removed.

File: Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/SpacePointMaker.hpp [hunk 2]
```diff
@@ -9,6 +9,13 @@
 #pragma once
 
 #include "Acts/Geometry/GeometryIdentifier.hpp"
+//#include "Acts/SpacePointFormation/DoubleHitSpacePointBuilder.hpp" TODO add
```
Comment (asalzburger) [inline Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/SpacePointMaker.hpp:12]: Somehow it would be nice to harmonise this to a single interface/module

File: Examples/Algorithms/TrackFinding/src/SpacePointMaker.cpp [hunk 3]
```diff
@@ -102,75 +107,18 @@ ActsExamples::ProcessCode ActsExamples::SpacePointMaker::execute(
 
     for (auto [moduleGeoId, moduleSourceLinks] : groupedByModule) {
       // find corresponding surface
-      const Acts::Surface* surface =
```
Comment (asalzburger) [inline Examples/Algorithms/TrackFinding/src/SpacePointMaker.cpp:105]: Where did the code from the lambda go?

File: Examples/Framework/include/ActsExamples/EventData/ModuleCluster.hpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#include "Acts/Digitization/Segmentation.hpp"
+#include "ActsExamples/EventData/IndexSourceLink.hpp"
+#include "ActsExamples/EventData/Measurement.hpp"
+
+#include <cassert>
+#include <vector>
+
+namespace ActsExamples {
+template <typename measurement_t>
+class ModuleCluster {
```
Comment (asalzburger) [inline Examples/Framework/include/ActsExamples/EventData/ModuleCluster.hpp:20]: Is the only need for this to forward the module segmentation ?
We could probably do this via the GeometryIdentifer ...

**Review by asalzburger — CHANGES_REQUESTED** [review]
I have added a few questions

File: Examples/Algorithms/TrackFinding/src/SpacePointMaker.cpp (hunk 3 above)
Comment (toyamaza) [reply Examples/Algorithms/TrackFinding/src/SpacePointMaker.cpp:105]: Thanks for the comments, Andi. What is the code from the lambda?

File: Examples/Framework/include/ActsExamples/EventData/ModuleCluster.hpp (hunk 4 above)
Comment (toyamaza) [reply Examples/Framework/include/ActsExamples/EventData/ModuleCluster.hpp:20]: Are you suggesting using Measurement as input for both pixel and strip SP builders, retrieving the module segmentation via GeometryIdentifier in the strip SP builder?

File: Core/include/Acts/SpacePointFormation/SpacePointBuilderConfig.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  double stripLengthTolerance = 0.01;
+  /// Allowed increase of strip length wrt gaps between strips
+  double stripLengthGapTolerance = 0.01;
+  /// Assumed position of the vertex
+  Vector3 vertex = {0., 0., 0.};
+  /// Perform the perpendicular projection for space point finding
+  bool usePerpProj = false;
+
+  DoubleHitSpacePointBuilderConfig()=default;
+};
```
Comment (CarloVarni) [inline Core/include/Acts/SpacePointFormation/SpacePointBuilderConfig.h:42]: Hi @toyamaza, why are these here but you also have a `Config` struct inside the SingleHitSpacePointBuilder class definition?

What about having a struct `SingleHitSpacePointBuilderConfig` with only a `std::shared_ptr<const Acts::TrackingGeometry> trackingGeometry; ` and then have `DoubleHitSpacePointBuilderConfig` inherit from it adding the remaining properties?

File: Core/include/Acts/SpacePointFormation/SingleHitSpacePointBuilder.hpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+  /// @brief Calculates the space points out of a given collection of clusters
+  /// and stores the results
+  ///
+  /// @param gctx The current geometry context object, e.g. alignment
+  /// @param clusters vector of clusters
+  /// @param spacePointStorage storage of the results
+  void calculateSpacePoints(const GeometryContext& gctx,
+                            const std::vector<cluster_t>& clusters,
+                            std::vector<spacepoint_t>& spacePointStorage) const;
```
Comment (CarloVarni) [inline Core/include/Acts/SpacePointFormation/SingleHitSpacePointBuilder.hpp:46]: What about using iterators for the input collection and std::back_insert_iterator for the output collection? Would this make sense?

Something such as 
```
template<template <typename...> typename container_t, typename iterator_t>
void calculateSpacePoints(const GeometryContext& gctx,
                            const iterator_t& clusters_begin,
                            const iterator_t& clusters_end,
                            std::back_insert_iterator<container_t<spacepoint_t>>& spacePointStorage) const;
```

File: Core/include/Acts/SpacePointFormation/DoubleHitSpacePointBuilder.hpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  ///
+  /// @param gctx The current geometry context object, e.g. alignment
+  /// @param clusterPairs pairs of clusters that are space point
+  /// candidates
+  /// @param spacePoints storage of the results
+  void calculateSpacePoints(
+      const GeometryContext& gctx,
+      const std::vector<std::pair<const cluster_t*, const cluster_t*>>&
+          clusterPairs,
+      std::vector<spacepoint_t>& spacePoints) const;
```
Comment (CarloVarni) [inline Core/include/Acts/SpacePointFormation/DoubleHitSpacePointBuilder.hpp:68]: what about introducing also a `calculateSpacePoints` method with the same inputs as in the SingleHitSpacePointBuilder class (thus moving towards a common interface), which internally calls makeClusterPairs and this method? Is this feasible?

```
void calculateSpacePoints(const GeometryContext& gctx,
                            const std::vector<cluster_t>& clusters,
                            std::vector<spacepoint_t>& spacePointStorage) const;
```

File: Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  for (const auto& clus : clusters) {
+    auto measurement = clus.measurement();
+    auto [gPos, gCov] = globalCoords(gctx, clus);
+    auto slink =
+        std::visit([](const auto& x) { return x.sourceLink(); }, measurement);
+
+    auto sp = spacepoint_t(gPos, gCov[0], gCov[1], {slink.index()});
+
+    spacePointStorage.emplace_back(sp);
+  }
```
Comment (CarloVarni) [inline Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp:99]: What about defining a new public method `calculateSpacePoint` that only accepts one cluster and produce one space point? In this was you can call that method here inside the loop and we can easily adapt the code in case `cluster_t` is a `std::variant` (as currently is in Athena)?

Would this make sense to you?

File: Examples/Framework/include/ActsExamples/EventData/ModuleCluster.hpp [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+#pragma once
+
+#include "Acts/Digitization/Segmentation.hpp"
+#include "ActsExamples/EventData/IndexSourceLink.hpp"
+#include "ActsExamples/EventData/Measurement.hpp"
+
+#include <cassert>
+#include <vector>
+
+namespace ActsExamples {
```
Comment (CarloVarni) [inline Examples/Framework/include/ActsExamples/EventData/ModuleCluster.hpp:18]: Is there a corresponding class in Core or should customers create their own version of this in their code?

File: Examples/Framework/include/ActsExamples/EventData/SimSpacePoint.hpp [hunk 10]
```diff
@@ -58,15 +60,15 @@ class SimSpacePoint {
   // Variance in rho/z of the global coordinates
   float m_varianceRho;
   float m_varianceZ;
-  // Index of the corresponding measurement
-  Index m_measurementIndex;
+  // Indices of the corresponding measurements
+  std::vector<Index> m_measurementIndices;
```
Comment (CarloVarni) [inline Examples/Framework/include/ActsExamples/EventData/SimSpacePoint.hpp:64]: What about using `boost::container::static_vector<Index, 2>` instead of `std::vector<Index>`?

**Comment by CarloVarni** [comment]
Hi @toyamaza  Sorry to jump in this PR. I had a few comments on the code changes I hope we can discuss

File: Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                         const std::vector<cluster_t>& clusters,
+                         std::vector<spacepoint_t>& spacePointStorage) const {
+  // Set the space point for all stored hits
+  for (const auto& clus : clusters) {
+    auto measurement = clus.measurement();
+    auto [gPos, gCov] = globalCoords(gctx, clus);
+    auto slink =
+        std::visit([](const auto& x) { return x.sourceLink(); }, measurement);
+
+    auto sp = spacepoint_t(gPos, gCov[0], gCov[1], {slink.index()});
```
Comment (CarloVarni) [inline Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp:96]: This assumes the space point class constructors. How should we handle ad-hoc constructors from the customers? Would this imply we'll have to create a space point collection with an acts-class and then convert (or copy) that to any customer-defined space point class?

Comment (CarloVarni) [reply Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp:96]: maybe this is more a question for @paulgessinger and @asalzburger since is more a general question also affecting other pieces of track reconstruction?

Comment (paulgessinger) [reply Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp:96]: Hm, good question. I guess this kind of depends on whether we keep the separation between internal and external spacepoints.

**Comment by paulgessinger** [comment]
Is there any way this PR could be broken up? For example I could imagine making the reorganization one PR and then the actual changes to the space point builders in a separate one. 

It's a bit difficult to determine what the changes were that were made on top of the moved files.

**Comment by toyamaza** [comment]
@paulgessinger, that makes sense. I will open two PRs for the reorganization and the SP builder changes.

File: Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp (hunk 11 above)
Comment (paulgessinger) [reply Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp:96]: There's a similar pattern with `std::function` used in the vertexing to convert input tracks, if I remember correctly. One concern is of course the potential virtual call overhead, although that might not be dominant here (and therefore not matter).

**Comment by toyamaza** [comment]
Let me close this PR as I split this into two PRs. The comments on the SP builder update above can be addressed/discussed in another PR.
